---
title: "PCAngsd"
author: "J. Ignacio Lucas LledÃ³"
date: "8/12/2021"
output: html_document
bibliography: references.bib
---

Here I will use PCAngsd [@Meisner2018] and possibly also NGSadmix [@Skotte2013]
to determine population structure. Both require genotype likelihoods for all
individuals. I can generate them directly from BAM files with ANGSD.

Because we already know that *Hemiechinus* and *Atelerix* are different genera,
I'd rather exclude them from the analysis, so that we focus on the *Erinaceus*
subjects.

```{bash fromBAM}
if [ ! -e popfile.txt ]; then
   if [ ! -e make_popfile.sh ]; then cp ../2021-12-06/make_popfile.sh ./; fi
   ./make_popfile.sh
fi
if [ ! -e lowest15.txt ]; then
   cp ../2021-12-01/lowest15.txt ./
fi
if [ ! -e bestErinaceus.txt ]; then
   grep -v -f lowest15.txt popfile.txt | \
   grep -v -E "(atelerix|hemiechinus)" | cut -f 1 > bestErinaceus.txt
fi

# I will use only forward reads, to save time and memory
BAMDIR='../2021-11-22'
PCANGSD=~/bin/pcangsd/pcangsd.py

if [ ! -e bam.filelist ]; then
   ls -1 $BAMDIR/*.bam | grep -f bestErinaceus.txt > bam.filelist
fi

if [ ! -e all.beagle.gz ]; then
   angsd -GL 2 -minInd 57 -doGlf 2 -minMapQ 30 -minQ 20 -setMinDepthInd 5 \
         -doMaf 1 -ref ../../data/reference.fa -SNP_pval 1.0e-6 -rmTriallelic 0.05 \
         -doMajorMinor 1 \
         -out all -nThreads 40 -bam bam.filelist 1> all.beagle.log 2> all.beagle.err
fi

if [ ! -e all.cov ]; then
   python $PCANGSD -beagle all.beagle.gz \
                   -minMaf 0.05 \
                   -threads 40 \
                   -out all 1> all.cov.log 2> all.cov.err
fi

# Expecting K=6: 2 europaeus, 2 roumanicus, 1 concolor, 1 amurensis.
for k in 3 4 5 6 7 8; do
   if [ ! -e K$k.qopt ]; then
      NGSadmix -likes all.beagle.gz \
               -K $k \
               -minMaf 0.02 \
               -method 0 \
               -outfiles K$k \
               -P 40 1> K$k.ngsadmix.log 2> K$k.ngsadmix.err
   fi
done
```

I assume that the output of both PCAngsd and NgsAdmix keep the order of the
samples as in the beagle-formated input file, which in turn should be the order
in which BAM files were specified.

```{r PCA}
pop <- read.table('popfile.txt', col.names = c('ind','pop'),
                  colClasses = c('character', 'factor'), row.names = 1)
bamlist <- scan('bam.filelist', what = 'character')
sampleList <- sapply(bamlist, function(x) {
  sub('.bam', '', unlist(strsplit(x, '/', fixed = TRUE))[3], fixed = TRUE) 
}, USE.NAMES = FALSE)

Erinaceus <- pop[sampleList, , drop = FALSE]
Erinaceus$pop <- droplevels(Erinaceus$pop)
byPop <- order(Erinaceus$pop)

COV <- as.matrix(read.table('all.cov'))
E   <- eigen(COV)
barplot(E$values[1:5] / sum(E$values), ylab = 'Explained variance', xlab = 'PC')
plot(E$vectors[, 1:2], col = Erinaceus$pop, xlab = 'PC 1', ylab = 'PC 2')
legend('topright', fill = 1:6, levels(Erinaceus$pop))
```

```{r NgsAdmix, fig.width=10}
library('plyr')
library('RColorBrewer')
# This was to help ordering samples for a better display. Not used.
JoinClosest  <- function(Q){
   n <- ncol(Q)
   Distances <- data.frame(
      Binary = as.numeric(dist(t(round(Q, 3) > 0), method = 'binary')),
      Group1 = unlist(sapply(2:n, function(x) seq(x, n))),
      Group2 = unlist(mapply(rep, 1:(n-1), (n-1):1))
   )
   if (any(round(Distances$Binary, 2) < 1) & n > 2) {
      Distances <- Distances[order(Distances$Binary),]
      closest   <- c(Distances[1,2], Distances[1,3])
      Q2 <- cbind(rowSums(Q[, closest]), Q[, -closest])
      colnames(Q2)[1] <- paste0(colnames(Q)[closest], collapse='.')
   } else {
      Q2 <- Q
   }
   return(Q2)
}

Q <- read.table('K6.qopt')
#row.names(Q) <- sampleList
row.names(Q) <- paste(sampleList, ' (',
                      substring(Erinaceus[sampleList, 'pop'],1,3),
                      '?)', sep = '')
Pop2Column <- ddply(Q, .(pop[sampleList, 'pop']), colSums)
names(Pop2Column)[1] <- 'pop'
Pop2Column$pop <- as.character(Pop2Column$pop)
z <- Pop2Column[sapply(Pop2Column[,2:7], which.max), 'pop']
colnames(Q) <- paste(z, duplicated(z)+1, sep = '_')
head(Q)

z <- order(round(Q[, 'europaeus_1'], 2),
           round(Q[, 'europaeus_2'], 2),
           round(Q[, 'roumanicus_1'], 2),
           round(Q[, 'roumanicus_2'], 2),
           round(Q[, 'concolor_1'], 2), decreasing = TRUE)
barplot(t(Q[z,]), col = brewer.pal(6, 'Accent'), border = NA,
        xlab = '', ylab = 'Admixture proportions for K=6', las=2,
        cex.names=0.5)
```

## New population assignment

Below I create a new population assignment file.

```{r newPop}
NewPop <- sapply(1:dim(Q)[1], function(x) {
   z <- order(Q[x,], decreasing = TRUE)
   q <- round(Q[x, z], 2)
   p <- colnames(Q)[z]
   f <- q > 0
   paste(p[f], q[f], sep=':', collapse=',')
})
names(NewPop) <- sampleList
pop$ngsadmix <- NewPop[row.names(pop)]
write.table(pop, file = 'popfile2.txt', quote = FALSE, sep = '\t')
```

With the new population assignment I can select individuals supposed to
have a single ancestry, and then calculate site-frequency spectra only with
those.

```{r sessionInfo}
sessionInfo()
```

## References

